### ISSUE TITLE
Blazor WASM with AAD groups and roles

### ISSUE THREAD

**Issue by guardrex** [dotnet/AspNetCore.Docs#17683]
In progress.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 217bc7a2-94db-e23b-9e74-7ad48b884462
* Version Independent ID: 071763aa-0173-af10-9634-9c50a4d5868b
* Content: [ASP.NET Core Blazor WebAssembly with Azure Active Directory groups and roles](https://docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/azure-active-directory-groups-and-roles?view=aspnetcore-5.0)
* Content Source: [aspnetcore/blazor/security/webassembly/azure-active-directory-groups-and-roles.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/security/webassembly/azure-active-directory-groups-and-roles.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by guardrex** [comment]
@mkArtakMSFT I hit a wall using the current approaches for this scenario. I'm going to put this issue on HOLD until we get past 3.2. Perhaps, Javier can comment on the scenario at that time and give me a tip to proceed.

For B2C to supply Azure AD security groups and roles data in order to create claims, the app **_has to make a separate Graph API call_**. I hoped that the setup would be similar to AAD ...

* The app is given Graph API [Directory.Read.All](https://docs.microsoft.com/en-us/graph/permissions-reference#directory-permissions) permission with admin consent in the portal. This is the least privileged permission for [List memberOf](https://docs.microsoft.com/en-us/graph/api/user-list-memberof). Ultimately (with a token), the request will be made to `/users/{id | userPrincipalName}/memberOf`. I just want to get a token tho to get started with this scenario.

* Add a scope ...

  ```csharp
  options.ProviderOptions.AdditionalScopesToConsent.Add(
      "https://graph.microsoft.com/Directory.Read.All");
  ```

  btw - I also tried using a `v1.0` segment here and in the token request. No dice.

* Ask for a token using the scope ...

  ```csharp
  var tokenResult = await TokenProvider.RequestAccessToken(
      new AccessTokenRequestOptions
      {
          Scopes = new[] { "https://graph.microsoft.com/Directory.Read.All" }
      });

  if (tokenResult.TryGetToken(out var token))
  {
      ...
  }
  else
  {
      ...
  }
  ```

  btw - I also tried the named client approach with scopes specified to `ConfigureHandler`. No dice.

It makes the request with ...

https://{ORGANIZATION}.b2clogin.com/{ORGANIZATION}.onmicrosoft.com/b2c_1_signupsignin1/oauth2/v2.0/authorize?response_type=token&scope=https%3A%2F%2Fgraph.microsoft.com%2FDirectory.Read.All%20openid%20profile&client_id={CLIENT ID}&redirect_uri=https%3A%2F%2Flocalhost%3A5001%2Fauthentication%2Flogin-callback&state={GUID}&nonce={GUID}&client_info=1&x-client-SKU=MSAL.JS&x-client-Ver=1.2.1&login_req={GUID}-b2c_1_signupsignin1&domain_req={GUID}&domain_hint=organizations&client-request-id={GUID}&prompt=none&response_mode=fragment

Azure B2C returns the redirect to the `authentication/login-callback` with ...

Location: https://localhost:5001/authentication/login-callback#error=interaction_required&error_description=AADB2C90077%3a+User+does+not+have+an+existing+session+and+request+prompt+parameter+has+a+value+of+%27None%27. [Correlation ID and Timestamp removed]

> User does not have an existing session and request prompt parameter has a value of 'None'.

... I've read from online sources that if the session (cookie) is missing, is misconfigured, or the session is expired then the silent token acquisition fails like this.

**Comment by mkArtakMSFT** [comment]
@javiercn can you unblock @guardrex here?

**Comment by javiercn** [comment]
I have no idea.

I've definitely been able to get a token to talk to Graph following those steps. It is likely that something is missing in the app configuration, but I don't know what.

I would suggest to start the flow within an in-private window?

**Comment by guardrex** [comment]
I'll hack around with it a bit more and see if I can work it out.

Yes on the in-private window. It has been so helpful that I added it to the guidance of our WASM security topics ...

https://docs.microsoft.com/en-us/aspnet/core/security/blazor/webassembly/standalone-with-authentication-library?view=aspnetcore-3.1#cookies-and-site-data

**Comment by guardrex** [comment]
**UPDATE**: A working prototype of the updates required for 5.0/Identity v2.0 is being discussed internally with framework engineers and will be placed on a PR to fix this issue. At the current pace, I suspect the PR will be up sometime between Monday, Nov. 23 and Wednesday, November 25.
